### PR TITLE
stats: add new TextReadout stat type

### LINKED ISSUE #5790: stats: New type(s) of Metric
Gauges are currently being repurposed for not-particularly-gaugey things, like [hashes](https://github.com/envoyproxy/envoy/blob/4e0b252a0b642e8ac669170130e2202244a53320/source/common/config/grpc_mux_subscription_impl.h#L58) and [booleans](https://github.com/envoyproxy/envoy/blob/4e0b252a0b642e8ac669170130e2202244a53320/source/common/config/grpc_stream.h#L61).

In #4974 I proposed separating these out into a new, more intentionally opaque, stats type. In #5770, @mattklein123 further proposed another type for text strings. This issue will be for 1) discussion of whether to do this and 2) tracking the work if so.

The current stats, which have some significant optimizations applied to their implementations, are all int64s. Does anyone who is familiar with the stats implementation immediately see any likely issues with introducing stats that aren't int64s? Would these strings have to have a maximum, or even fixed length?

Also, let's start the discussion of what to call them. I think it would be fun and useful to keep the physical control panel metaphor of "gauge" going, so I suggest Readout for the text one. For the opaque one, though, I think by its nature can't have a particularly good name, so I suggest just Opaque.

@ggreenway @jmarantz @mattklein123 @ramaraochavali 

(Also: I am considering trying to remove set() from Gauge, or at least making it test-only. However, I haven't yet inspected all of the Gauge::set() users closely enough to know for sure if it would be reasonable.)

### PR DESCRIPTION
**Pull request by efimki**
Description: Add the TextReadout stat type, for holding  strings. 
Currently ignored (not preserved) during hot restart. 
If importing text readout is desired in the future that could be a mode-bit like gauge.

Risk Level: low
Testing: roughly duplicated every bit of test logic involving the word "gauge".

Fixes #5790 @fredlas @jmarantz

### CODE REVIEW

File: source/common/stats/allocator_impl.cc [hunk 1]
```diff
@@ -228,6 +228,25 @@ class GaugeImpl : public StatsSharedImpl<Gauge> {
...
+    const size_t count = alloc_.text_readouts_.erase(statName());
+    ASSERT(count == 1);
+  }
+
+  // Stats::TextReadout
+  void set(const std::string& value) override { value_ = value; }
+  std::string value() const override { return value_; }
+
+private:
+  std::string value_;
```
Comment (jmarantz) [inline source/common/stats/allocator_impl.cc:247]: quick drive-by: I think value_ needs to be atomic.

Do you want to have an 8-byte readout that uses an atomic<uint64_t>? Or mutex-protect the value?

Also you might be sharing some bits with counter/gauge that you don't need by inheriting from StatsSharedImpl -- can you check?

Comment (jmarantz) [reply source/common/stats/allocator_impl.cc:247]: If you are going to manage TextReadout via the Allocator, yes. The alternative is to treat them more like histograms, which has this disadvantage that if you create a histogram by the same name twice, you'll get two distinct objects. This results in some confusing output in the admin /stats panel.

So I'd go with ref-counting; you could just copy the code. I think the shared class between counters and gauges steals some bits leftover from flags() for the refcount, and you wouldn't be able to do that, but it's not really a big deal if you are paying for a std::string and mutex anyway.

Comment (efimki) [reply source/common/stats/allocator_impl.cc:247]: Thanks for the suggestion! 

I've tried to extract RefcountMetricImpl base class out of StatsSharedImpl, but in this case StatsSharedImpl only contains std::atomic<uint64_t> value_{0} which seems silly.

I've extracted value_ into Counter and Gauge instead and AFAICT there is nothing else in StatsSharedImpl that is unused by TextReadout.

[I've made the change locally, but having troubles with `git push origin textstats` claiming that "Everything up-to-date"]

Comment (jmarantz) [reply source/common/stats/allocator_impl.cc:247]: I'm not the best person to advise you on git-fu; I am barely alive on the platform :)

However I finally just pulled up the code and I think you can just move the value_ field out of StatsSharedImpl and just put it in both CounterImpl and GaugeImpl. value_ is not ref'd directly by anything in StatsSharedImpl anyway. You'll need flags_ because used() is probably still important.

Then you can use StatsSharedImpl from all 3 stat variants, and not worry about ref-counting.

Comment (efimki) [reply source/common/stats/allocator_impl.cc:247]: Perfect, that's exactly what I did. :) I'll ping you back once I manage to upload my changes.

Comment (efimki) [reply source/common/stats/allocator_impl.cc:254]: I've managed to push my changes, which is good, but somehow picked up unrelated commits, which seems bad. 

I'll be glad to fix that if you have any suggestions or try to create another PR from a clean branch.

Comment (efimki) [reply source/common/stats/allocator_impl.cc:254]: It looks like I've been able to get rid of unrelated commits, so PTAL when you have a chance.

File: include/envoy/stats/scope.h [hunk 2]
```diff
@@ -17,12 +17,14 @@ namespace Stats {
 class Counter;
 class Gauge;
 class Histogram;
+class TextReadout;
```
Comment (jmarantz) [inline include/envoy/stats/scope.h:20]: abc order (while you are in here fix NullGaugeImpl as well)

File: source/common/stats/allocator_impl.cc [hunk 3]
```diff
@@ -226,6 +223,35 @@ class GaugeImpl : public StatsSharedImpl<Gauge> {
...
+    Thread::LockGuard lock(mutex_);
+    value_ = value;
+  }
+  std::string value() const override {
+    Thread::LockGuard lock(mutex_);
+    return value_;
+  }
+
+private:
+  mutable Thread::MutexBasicLockable mutex_;
```
Comment (jmarantz) [inline source/common/stats/allocator_impl.cc:253]: can you make this an absl::Mutex directly? No need for this particular mutex to implement a pure interface, to be compatible with shared-memory. I think that will save some bytes per TextReadout.

File: source/common/stats/allocator_impl.cc [hunk 4]
```diff
@@ -256,6 +282,19 @@ GaugeSharedPtr AllocatorImpl::makeGauge(StatName name, StatName tag_extracted_na
   return gauge;
 }
 
+TextReadoutSharedPtr AllocatorImpl::makeTextReadout(StatName name, StatName tag_extracted_name,
+                                                    const StatNameTagVector& stat_name_tags) {
+  Thread::LockGuard lock(mutex_);
+  ASSERT(gauges_.find(name) == gauges_.end());
```
Comment (jmarantz) [inline source/common/stats/allocator_impl.cc:288]: also assert there's no counter of the same name? And in the makeCounter and makGauge impls make sure there are no TextReadouts of the same name.

File: source/common/stats/isolated_store_impl.h [hunk 5]
```diff
@@ -68,6 +71,17 @@ template <class Base> class IsolatedStatsCache {
     return *new_stat;
   }
 
+  Base& get(StatName name, bool b) {
```
Comment (jmarantz) [inline source/common/stats/isolated_store_impl.h:74]: s/b/lower_case/ ? but I have some questions:

1. do we need lower-casing in the stat implementation? Or should whatever value transforms are needed be done at call-site?
2. I don't see lower-casing in the allocator_impl.cc version.

**Comment by jmarantz** [comment]
More thoughts:
 * how should text-readout be displayed in the admin /stats endpoint? Please be sure to properly html-escape output from the text, to avoid xss (or just weird formatting on the output)
 * how should text-readout be send to prometheus?
 * it's probably worth auditing all calls to counters() and gauges() in source/... to see if we need to have similar handling for text-readouts.

**Comment by htuch** [comment]
Please merge master to pick up https://github.com/envoyproxy/envoy/pull/10672. We no longer accept changes to v2 (without explicit exception), so any API modifications should happen in v3. If this PR is adding a new proto, please follow the updated instructions in https://github.com/envoyproxy/envoy/blob/master/api/STYLE.md#adding-an-extension-configuration-to-the-api.

File: source/common/stats/isolated_store_impl.h (hunk 5 above)
Comment (efimki) [reply source/common/stats/isolated_store_impl.h:74]: Ah, you've found my hack. :) 

The problem is that there are multiple overloads of get() and Base constructor. I've added a dummy bool parameter to disambiguate text readouts from counters. I've initially named it |lower_case| as I thought that it could be useful, but I'm not convinced that it is, so I've named it |b| here.

Comment (jmarantz) [reply source/common/stats/isolated_store_impl.h:74]: I suspected that :)

Can you just add an enum to disambiguate in a self-documenting way?

**Comment by jmarantz** [comment]
Simply not importing them during hot restart is fine for now. You could leave a comment staying that if importing text-readout is desired in the future that could be a mode-bit like gauge, and that could be added later.

File: source/common/stats/isolated_store_impl.h (hunk 5 above)
Comment (efimki) [reply source/common/stats/isolated_store_impl.h]: Done

**Comment by efimki** [comment]
FYI, I think "efimki requested review from alyssawilk, htuch, lizan, PiotrSikora and snowp as code owners yesterday" was auto-requested by repokitteh bot that noticed some spurious commits pulled from upstream and unintentionally included into this PR due to my github workflow clumsiness.

**Comment by jmarantz** [comment]
Looks good to me generally, but:
 * you need to do a master merge to resolve stats_Integration_test. when you do that, you should manually re-test with "-c opt  --test_env=ENVOY_MEMORY_TEST_EXACT=true" to iterate toward the correct regold.
 * we need to decide what to do about admin /stats and audit other uses of `.counters()` and `.gauges()`

**Comment by efimki** [comment]
@jmarantz - thanks for your review and suggestions. 
I've re-run the test with "-c opt --test_env=ENVOY_MEMORY_TEST_EXACT=true" and updated numbers.

With regards to escaping text readout values is there a common pattern that I could follow?

I'm wondering whether it makes sense to replace .value() getter with safeHtmlEscapedValue() and unsafeRawStringValue() getters to make it clear that stored value is not escaped. 

I think admin /stats will NOT be affected by this PR as they have to explicitly iterate TextReadout stats. I'll be happy to modify admin/stats in this PR, but I'm not sure whether it is the right thing to do. WDYT?

**Comment by jmarantz** [comment]
I'm OK with a follow-up to add visibility to the stats. But would like to definitely have a follow-up, as generally developers in Envoy that see an option for a new stat would expect to be able to get at the value through the UI or their favorite stats monitoring system (e.g. prometheus). Not sure what support those systems would have for a text-readout type. But at the minimum it should be an admin /stats endpoint.

It's a fair argument that you don't need to become an effort in every stats sink to add a new stats type, but you should definitely sprinkle some TODOs around for others to note that there is data that is collected and not exposed.

I don't think it's necessary for you to handle the HTML escaping or rename the methods in the stats type. HTML is just one of many possible output formats that might need some kind of escaping, another being JSON. You don't need to comment about all possible outputs in the value() method here.

**Comment by htuch** [comment]
FWIW, I think this PR would be great for xDS version readout, I've filed https://github.com/envoyproxy/envoy/issues/10736.

File: source/common/stats/allocator_impl.cc [hunk 6]
```diff
@@ -226,12 +223,42 @@ class GaugeImpl : public StatsSharedImpl<Gauge> {
...
+      : StatsSharedImpl(name, alloc, tag_extracted_name, stat_name_tags) {}
+
+  void removeFromSetLockHeld() EXCLUSIVE_LOCKS_REQUIRED(alloc_.mutex_) override {
+    const size_t count = alloc_.text_readouts_.erase(statName());
+    ASSERT(count == 1);
+  }
+
+  // Stats::TextReadout
+  void set(const std::string& value) override {
+    absl::MutexLock lock(&mutex_);
```
[the remark this answers is not in the document]
Comment (efimki) [reply source/common/stats/allocator_impl.cc:244]: Great suggestion! Done.

File: source/common/stats/allocator_impl.cc [hunk 7]
```diff
@@ -226,12 +223,42 @@ class GaugeImpl : public StatsSharedImpl<Gauge> {
...
+    value_ = value;
+  }
+  std::string value() const override {
+    absl::MutexLock lock(&mutex_);
+    return value_;
+  }
+
+private:
+  mutable absl::Mutex mutex_;
+  std::string value_;
```
[the remark this answers is not in the document]
Comment (efimki) [reply source/common/stats/allocator_impl.cc:254]: Good suggestion, Done.

File: include/envoy/stats/stats.h [hunk 8]
```diff
@@ -156,5 +156,25 @@ class Gauge : public Metric {
...
+  // Text readout type is used internally to disambiguate isolated store
+  // constructors. In the future we can extend it to specify text encoding or
+  // some such.
+  enum class Type {
+    Default, // No particular meaning.
+  };
+
+  virtual ~TextReadout() {}
+
+  virtual void set(std::string&& value) PURE;
```
Comment (jmarantz) [inline include/envoy/stats/stats.h:173]: Needs doc, especially since it's got a potentially unexpected (though justified) behavior. In fact I might even call this 'move' rather than 'set'?

doc also for the accessor.

File: source/common/stats/allocator_impl.cc (hunk 6 above)
[the remark this answers is not in the document]
Comment (jmarantz) [reply source/common/stats/allocator_impl.cc:244]: actually I just thought of another idea; may be this would be more intuitive?

Leave this API as is, but change the implementation to copy the input param to a local without taking the lock, and then std::move it into the structure after taking the lock.

[the remark this answers is not in the document]
Comment (efimki) [reply source/common/stats/allocator_impl.cc:244]: I agree that having `set()` method move input value is quite surprising, and I was going to suggest to keep `const std::string&` and add `swap(std::string& value)` method, but I did find quite a few methods that take `&&`, so I've scraped that. :)

To be honest I would argue that lock contention is not an issue as I don't expect to have text readout values changed racily from multiple threads. 
I might be coming from a very narrow use case point of view, but in that use case the value of the text readout is unlikely to change ever.
What use case are you thinking about?

File: include/envoy/stats/stats.h (hunk 8 above)
Comment (efimki) [reply include/envoy/stats/stats.h:177]: Done.

**Comment by efimki** [comment]
Hrm, AFAICT envoy-presubmit test fails [1] but some memory allocation expectations don't match. Strangely those are numbers that I get when I run it locally [2].
What is a good way to address this? [3]
Should I use the numbers observed by envoy-presubmit and ignore local numbers?

1. https://dev.azure.com/cncf/envoy/_build/results?buildId=37226&view=ms.vss-test-web.build-test-results-tab&runId=288850&resultId=111919&paneView=debug

2. bazel test -c opt --test_env=ENVOY_MEMORY_TEST_EXACT=true       test/integration:stats_integration_test

3. https://github.com/envoyproxy/envoy/blob/master/source/docs/stats.md#stats-memory-tests

**Comment by jmarantz** [comment]
(a) yes: use the presubmit numbers 
(b) also figure out why your numbers are different. Are you using the same version of clang and the c++ std library?

**Comment by efimki** [comment]
@jmarantz - good call about comparing compiler versions! It looks like circle-ci is using `clang 9.0.0` but my machine has `clang 8.0.1`. 

I don't see the `stats_integration_test` failures any more, which is good. Unfortunately I still see some cryptic error from envoy-presubmit (Linux-x64 release) that couldn't find "advapi32.dll". 
Missing dll doesn't surprise me on linux64, but I'm surprised that it is actually needed. 

```
##[warning]Failed to upload file TestResults_289018.zip to Blob , exception message System.DllNotFoundException: Unable to load shared library 'advapi32.dll' or one of its dependencies. In order to help diagnose loading problems, consider setting the LD_DEBUG environment variable: libadvapi32.dll: cannot open shared object file: No such file or directory
   at Microsoft.TeamFoundation.TestClient.PublishTestResults.TestLogStore._executeWithRetry(Func`1 action, Func`1 actionWithRename, Func`1 actionOnSasExpiry, Boolean allowDuplicate, Boolean allowRetryOnNetworkingError, TelemetryTracker telemetryTracker)
   at Microsoft.TeamFoundation.TestClient.PublishTestResults.TestLogStore.UploadInternalAsync(Guid projectId, TestLogScope scope, Int32 buildId, Int32 runId, String logFileSourcePath, Stream stream, String fileName, TestLogType logType, Int32 testResultId, Int32 testSubResultId, Dictionary`2 metaData, TelemetryTracker telTracker, Boolean allowDuplicate, TestLogCompressionType logCompressionType, String originalFileContentType)
```

**Comment by jmarantz** [comment]
our old friend win32 :)  I'm not clear on what CI magic causes that to show up in linux-x64 release. @lizan does this ring a bell?

File: include/envoy/stats/allocator.h [hunk 9]
```diff
@@ -49,6 +49,15 @@ class Allocator {
                                    const StatNameTagVector& stat_name_tags,
                                    Gauge::ImportMode import_mode) PURE;
 
+  /**
+   * @param name the full name of the stat.
+   * @param tag_extracted_name the name of the stat with tag-values stripped out.
+   * @param tags the tag values.
+   * @return TextReadoutSharedPtr a text readout, or nullptr if allocation failed, in which case
+   *     tag_extracted_name and tags are not moved.
```
Comment (mattklein123) [inline include/envoy/stats/allocator.h:57]: I don't think allocation can fail anymore (this is a holdover from the shared mem days), right? Can you fix this above also?

File: include/envoy/stats/scope.h [hunk 10]
```diff
@@ -136,6 +138,31 @@ class Scope {
...
 
+  /**
+   * Creates a TextReadout from the stat name. Tag extraction will be performed on the name.
+   * @param name The name of the stat, obtained from the SymbolTable.
+   * @return a text readout within the scope's namespace.
+   */
+  TextReadout& textReadoutFromStatName(const StatName& name) {
+    return textReadoutFromStatNameWithTags(name, absl::nullopt);
+  }
+  /**
```
Comment (mattklein123) [inline include/envoy/stats/scope.h:149]: nit: newline

File: include/envoy/stats/scope.h [hunk 11]
```diff
@@ -136,6 +138,31 @@ class Scope {
...
+   * will be performed on the name.
+   * @param name The name of the stat, obtained from the SymbolTable.
+   * @param tags optionally specified tags.
+   * @return a text readout within the scope's namespace.
+   */
+  virtual TextReadout& textReadoutFromStatNameWithTags(const StatName& name,
+                                                       StatNameTagVectorOptConstRef tags) PURE;
+
+  /**
+   * TODO(#6667): this variant is deprecated: use textReadoutFromStatName.
```
Comment (mattklein123) [inline include/envoy/stats/scope.h:160]: For new code I think we can remove this?

Comment (jmarantz) [reply include/envoy/stats/scope.h:160]: We still need this for stats macros to work. However we should put the format-check at https://github.com/envoyproxy/envoy/blob/0ebcd1a5ea904e849b62763bcede94a5adea7c60/tools/code_format/check_format.py#L670 in this PR.

File: include/envoy/stats/allocator.h (hunk 9 above)
Comment (mattklein123) [reply include/envoy/stats/allocator.h:57]: We crash on OOM so I think these comments can just be deleted. Thank you!

/wait

Comment (efimki) [reply include/envoy/stats/allocator.h]: Thanks for the confirmation! Done.

**Comment by efimki** [comment]
Hrm, the circleci failure seems unrelated and in-actionable. 
Any ideas how to fix this? 
Should I try to `kick-ci` again?

```
[ RUN      ] Protocols/BufferIntegrationTest.RouteOverride/IPv4_Http2Downstream_HttpUpstream
#0 Envoy::SignalAction::sigHandler() [0x10dd7cb4]
#1 __restore_rt [0x7fe30341d390]
#2 Envoy::BaseIntegrationTest::createEnvoy() [0xefa4ad7]
#3 Envoy::BaseIntegrationTest::initialize() [0xefa36ba]
#4 __llvm_coverage_mapping [0x9ad1263]
#5 testing::internal::HandleSehExceptionsInMethodIfSupported<>() [0x12215686]
#6 testing::internal::HandleExceptionsInMethodIfSupported<>() [0x12201c51]
#7 testing::Test::Run() [0x121ec522]
#8 testing::TestInfo::Run() [0x121ed028]
#9 testing::TestSuite::Run() [0x121ed799]
#10 testing::internal::UnitTestImpl::RunAllTests() [0x121fa7a6]
#11 testing::internal::HandleSehExceptionsInMethodIfSupported<>() [0x1221a526]
#12 testing::internal::HandleExceptionsInMethodIfSupported<>() [0x12204a71]
#13 testing::UnitTest::Run() [0x121fa1f0]
#14 RUN_ALL_TESTS() [0x1076b235]
#15 Envoy::TestRunner::RunTests() [0x1076a936]
#16 main [0x10768c67]
#17 __libc_start_main [0x7fe303062830]
external/bazel_tools/tools/test/collect_coverage.sh: line 131: 16404 Aborted                 (core dumped) "$@"

```

**Comment by mattklein123** [comment]
The clang-tidy errors are legit. Can you fix them? Coverage is a flake and we can rerun if needed. There are known issues there that are being worked on.

/wait

**Comment by efimki** [comment]
Hrm, "macOS was canceled" after 6 hours. 
What's a good next step? `kick-ci`?

**Comment by mattklein123** [comment]
Please push and empty commit to kick Circle CI, I don't know what is broken there.

**Comment by mattklein123** [comment]
Sigh CircleCI. I dont know what is broken. I will try closing and reopening. If that doesn't work you will probably need to make a fresh PR and we can get this merged. I just want to make sure API/docs/etc. are all passing.

**Comment by efimki** [comment]
Looks like MacOS pipeline has timed out after 6 hours, but AFAICT the executed tests have passed:

https://dev.azure.com/cncf/envoy/_build/results?buildId=37491&view=logs&j=a5e52b91-c83f-5429-4a68-c246fc63a4f7&t=5852bf5a-5a02-52f3-bee8-4fdc90cda9d0

/azp run

**Comment by efimki** [comment]
Woo-hoo, thanks Joshua!
